File: app/src/main/java/com/boardgamegeek/ui/GameLinksFragment.java
package com.boardgamegeek.ui;


import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.ColorInt;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.boardgamegeek.R;
import com.boardgamegeek.provider.BggContract;
import com.boardgamegeek.ui.GameActivity.ColorEvent;
import com.boardgamegeek.util.ActivityUtils;
import com.boardgamegeek.util.PaletteUtils;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

import java.util.List;

import butterknife.BindViews;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;
import hugo.weaving.DebugLog;

public class GameLinksFragment extends Fragment {
	private static final String KEY_GAME_ID = "GAME_ID";
	private static final String KEY_GAME_NAME = "GAME_NAME";
	private static final String KEY_ICON_COLOR = "ICON_COLOR";

	private int gameId;
	private String gameName;
	@ColorInt private int iconColor = Color.TRANSPARENT;

	Unbinder unbinder;
	@BindViews({
		R.id.icon_link_bgg,
		R.id.icon_link_bg_prices,
		R.id.icon_link_amazon,
		R.id.icon_link_ebay
	}) List<ImageView> colorizedIcons;

	public static GameLinksFragment newInstance(int gameId, String gameName, @ColorInt int iconColor) {
		Bundle args = new Bundle();
		args.putInt(KEY_GAME_ID, gameId);
		args.putString(KEY_GAME_NAME, gameName);
		args.putInt(KEY_ICON_COLOR, iconColor);
		GameLinksFragment fragment = new GameLinksFragment();
		fragment.setArguments(args);
		return fragment;
	}

	@Override
	public void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		EventBus.getDefault().register(this);
	}

	@DebugLog
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		View rootView = inflater.inflate(R.layout.fragment_game_links, container, false);
		unbinder = ButterKnife.bind(this, rootView);
		readBundle(getArguments());
		colorize();
		return rootView;
	}

	private void readBundle(@Nullable Bundle bundle) {
		if (bundle == null) return;
		gameId = bundle.getInt(KEY_GAME_ID, BggContract.INVALID_ID);
		gameName = bundle.getString(KEY_GAME_NAME);
		iconColor = bundle.getInt(KEY_ICON_COLOR, Color.TRANSPARENT);
	}

	@DebugLog
	@Override
	public void onDestroyView() {
		super.onDestroyView();
		if (unbinder != null) unbinder.unbind();
	}

	@Override
	public void onDestroy() {
		super.onDestroy();
		EventBus.getDefault().unregister(this);
	}

	@SuppressWarnings("unused")
	@DebugLog
	@OnClick({ R.id.link_bgg, R.id.link_bg_prices, R.id.link_amazon, R.id.link_amazon_uk, R.id.link_amazon_de, R.id.link_ebay })
	void onLinkClick(View view) {
		switch (view.getId()) {
			case R.id.link_bgg:
				ActivityUtils.linkBgg(getActivity(), gameId);
				break;
			case R.id.link_bg_prices:
				ActivityUtils.linkBgPrices(getActivity(), gameName);
				break;
			case R.id.link_amazon:
				ActivityUtils.linkAmazon(getActivity(), gameName, ActivityUtils.LINK_AMAZON_COM);
				break;
			case R.id.link_amazon_uk:
				ActivityUtils.linkAmazon(getActivity(), gameName, ActivityUtils.LINK_AMAZON_UK);
				break;
			case R.id.link_amazon_de:
				ActivityUtils.linkAmazon(getActivity(), gameName, ActivityUtils.LINK_AMAZON_DE);
				break;
			case R.id.link_ebay:
				ActivityUtils.linkEbay(getActivity(), gameName);
				break;
		}
	}

	@SuppressWarnings("unused")
	@Subscribe
	public void onEvent(ColorEvent event) {
		if (event.getGameId() == gameId) {
			iconColor = event.getIconColor();
			colorize();
		}
	}

	@DebugLog
	private void colorize() {
		if (!isAdded()) return;
		if (iconColor != Color.TRANSPARENT) {
			ButterKnife.apply(colorizedIcons, PaletteUtils.rgbIconSetter, iconColor);
		}
	}
}

File: app/src/main/java/com/boardgamegeek/ui/GameLinksFragment.kt
package com.boardgamegeek.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.painter.Painter
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.dimensionResource
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import com.boardgamegeek.R
import com.boardgamegeek.databinding.FragmentNestedComposeViewBinding
import com.boardgamegeek.extensions.*
import com.boardgamegeek.provider.BggContract
import com.boardgamegeek.ui.compose.EmptyContent
import com.boardgamegeek.ui.theme.BggAppTheme
import com.boardgamegeek.ui.viewmodel.GameViewModel
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class GameLinksFragment : Fragment() {
    private var _binding: FragmentNestedComposeViewBinding? = null
    private val binding get() = _binding!!

    @Suppress("RedundantNullableReturnType")
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = FragmentNestedComposeViewBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.composeView.setContent {
            val viewModel by activityViewModels<GameViewModel>()
            val game by viewModel.game.observeAsState()
            val paddingValues = PaddingValues(dimensionResource(R.dimen.material_margin_horizontal), 8.dp)
            if (game == null) {
                EmptyContent(
                    stringResource(R.string.empty_game),
                    painterResource(R.drawable.link_24px),
                    Modifier
                        .fillMaxSize()
                        .padding(paddingValues),
                )
            } else {
                game?.let {
                    GameLinks(
                        it.id,
                        it.name,
                        iconColor = it.iconColor,
                        modifier = Modifier.padding(paddingValues),
                    )
                }
            }
        }
    }
}

@OptIn(ExperimentalMaterial3ExpressiveApi::class)
@Composable
private fun GameLinks(
    gameId: Int,
    gameName: String,
    modifier: Modifier = Modifier,
    iconColor: Int = android.graphics.Color.TRANSPARENT,
) {
    val context = LocalContext.current
    val tint = if (iconColor == android.graphics.Color.TRANSPARENT)
        LocalContentColor.current
    else
        Color(iconColor)
    Column(
        modifier = modifier,
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        if (gameId != BggContract.INVALID_ID) {
            LinkButton(
                textResId = R.string.link_geekbuddy_analysis,
                painterResource(R.drawable.geekbuddy_24px),
                tint = tint,
                onClick = { context.linkToBgg("geekbuddy/analyze/thing", gameId) }
            )
            LinkButton(
                textResId = R.string.link_bgg,
                painterResource(R.drawable.open_in_browser_24px),
                tint = tint,
                onClick = { context.linkBgg(gameId) }
            )
        }
        if (gameName.isNotBlank()) {
            Text(
                stringResource(R.string.title_acquire),
                style = MaterialTheme.typography.titleMedium,
                modifier = Modifier.padding(top = 24.dp, bottom = 4.dp)
            )
            LinkButton(
                textResId = R.string.link_bg_prices,
                painterResource(R.drawable.shopping_cart_24px),
                tint = tint,
                onClick = { context.linkBgPrices(gameName) }
            )
            LinkButton(
                textResId = R.string.link_amazon,
                painterResource(R.drawable.amazon_24px),
                tint = tint,
                onClick = { context.linkAmazon(gameName, LINK_AMAZON_COM) }
            )
            LinkButton(
                textResId = R.string.link_amazon_uk,
                painterResource(R.drawable.amazon_24px),
                tint = tint,
                onClick = { context.linkAmazon(gameName, LINK_AMAZON_UK) }
            )
            LinkButton(
                textResId = R.string.link_amazon_de,
                painterResource(R.drawable.amazon_24px),
                tint = tint,
                onClick = { context.linkAmazon(gameName, LINK_AMAZON_DE) }
            )
            LinkButton(
                textResId = R.string.link_ebay,
                painterResource(R.drawable.ebay_24px),
                tint = tint,
                onClick = { context.linkEbay(gameName) }
            )
        }
    }
}

@OptIn(ExperimentalMaterial3ExpressiveApi::class)
@Composable
private fun LinkButton(
    textResId: Int,
    icon: Painter,
    tint: Color,
    modifier: Modifier = Modifier,
    onClick: () -> Unit = {},
) {
    val size = ButtonDefaults.MediumContainerHeight
    OutlinedButton(
        onClick = onClick,
        modifier = modifier.heightIn(size),
        shape = ButtonDefaults.shapesFor(size).shape,
        contentPadding = ButtonDefaults.contentPaddingFor(size, true),
    ) {
        Icon(
            icon,
            contentDescription = null,
            modifier = Modifier.size(ButtonDefaults.iconSizeFor(size)),
            tint = tint,
        )
        Spacer(Modifier.size(ButtonDefaults.iconSpacingFor(size)))
        Text(
            stringResource(textResId),
        )
    }
}

@Preview(showBackground = true)
@Composable
private fun GameLinksPreview() {
    BggAppTheme {
        GameLinks(13, "Terra Mystica", iconColor = -14667640)
    }
}